Demonstration Recorder

awjuliani opened the pull request:
Adds Demonstration Recorder feature.

## Specifics C#:
* `DemonstrationRecorder` (Class - Inherits from MonoBehavior) - Component which attaches to GameObject with Agent Component. 
    * `CreateDemonstrationFile()` - Creates a new demonstration file, and appends a brainparameters json object to it.
    * `WriteExperience()` - Takes an AgentInfo, converts it to a json object, and appends it to the current demonstration file.
* `Demonstration` (ScriptableObject) contains metadata about recorded demonstration.
* `DemonstrationImporter` (ScriptedImporter) contains logic to load `.demo` file as Unity asset.
## Specifics Python:
* Create static `DemonstrationLoader` class which takes file path of `.demo` file, and returns a filled buffer ready to use for training.
* Rename `BehavioralCloningTrainer` to `OnlineBCTrainer`.
* Create `OfflineBCTrainer` which uses `DemonstrationLoader` and `.demo` file for experience instead of environment.
## Demonstrations:
* File is of type `.demo`.
* First line is json object corresponding to `BrainParameters`
* Last line is json object corresponding to number of experiences and number of episodes.
* All inner lines are json objects corresponding to `AgentInfo`.

## Todo:
- [x] Verify design
- [x] Add python tests
- [x] Add Unity tests
- [x] Add Player Window GUI
- [x] Make Inspector GUI for `Demonstration` readonly.
- [x] Update documentation
- [x] Add demonstration icon
- [x] Printer brain parameters should contain all info.

awjuliani commented on UnitySDK/Assets/ML-Agents/Editor/DemonstrationImporter.cs:
```diff
@@ -0,0 +1,45 @@
+            string demonstrationName = Path.GetFileName(ctx.assetPath);
+
+            var inputType = Path.GetExtension(ctx.assetPath);
+            if (inputType == null)
```
udfheufheurfh

eshvk commented on ml-agents/mlagents/trainers/demo_loader.py:
```diff
@@ -0,0 +1,70 @@
+
+class DemonstrationLoader(object):
+    @staticmethod
+    def load(file_path, brain_name, sequence_length):
```
Is there a reason `load` has to be a method of a class? Python (unlike C#) allows functions to exist outside of classes. Seems to me that the indirection is unnecessary.

eshvk commented on ml-agents/mlagents/trainers/trainer.py:
```diff
@@ -19,14 +19,15 @@ class UnityTrainerException(UnityException):
         :param run_id: The identifier of the current run
         """
-        self.brain_name = brain_name
+        self.TRAINER_NAME = 'Base'
```
From what I understand this `TRAINER_NAME` is used mainly during logging an exception? If this is correct, can we use the `.__class__` method of the object to determine what class generated this and put that in the exception? I think this will ensure that any future refactors of the class won't lead to the two class name and `TRAINER_NAME` being robust to desynchronization.

vladimiroster commented on UnitySDK/Assets/ML-Agents/Editor/DemonstrationImporter.cs:
```diff
@@ -0,0 +1,45 @@
+    /// <summary>
+    /// Asset Importer used to parse demonstration files.
+    /// </summary>
+    [ScriptedImporter(1, new[] {"demo"})]
```
"demo" is very general. Does it make sense to have a more unique extension?

awjuliani replied:
Any suggestions? `.demon`? 👿

vladimiroster commented on UnitySDK/Assets/ML-Agents/Editor/DemonstrationImporter.cs:
```diff
@@ -0,0 +1,45 @@
+            
+
+#if UNITY_2017_3_OR_NEWER
+            ctx.AddObjectToAsset(ctx.assetPath, demonstration);
```
Since this is a nice custom asset, perhaps an icon is in order

awjuliani replied:
Definitely. Going to add that to the to-do.

vladimiroster commented on UnitySDK/Assets/ML-Agents/Scripts/DemonstrationStore.cs:
```diff
@@ -0,0 +1,98 @@
+        public int API;
+    }
+    
+    public class DemonstrationStore
```
Documentation

vladimiroster commented on UnitySDK/Assets/ML-Agents/Scripts/DemonstrationStore.cs:
```diff
@@ -0,0 +1,98 @@
+        public void CreateDemonstrationFile(string demonstrationName)
+        {
+            // Creates demonstration file.
+            var literalName = demonstrationName;
```
Might make sense to sanitize this string for security

awjuliani replied:
Would stripping all characters except alphanumeric work?

vladimiroster replied:
i think so. the low hanging vulnerability here is demonstration name "../../../../some_important_file_i_dont_want_deleted"

awjuliani commented:
Thanks for the feedback, @eshvk and @mmattar. It is really helpful. I've addressed all of your comments, either fully or partially, so please feel free to take another look and provide another round of feedback. 

In particular, @mmattar, it would be helpful to get more feedback on the `DemonstrationStore`. I've moved the logic to a separate class, but it does not support modification of demonstrations, and as such each session corresponds to one file. Happy to explore ways to make this more general in the future.

mmattar commented on UnitySDK/Assets/ML-Agents/Scripts/DemonstrationStore.cs:
```diff
@@ -0,0 +1,103 @@
+        public int API;
+    }
+    
+    public class DemonstrationStore
```
The current API exposed by DemonstrationStore exposes several implementation details (e.g. creating a directory, managing a file and saving metadata). It should really only contain 3 methods to abstract away that under the hood it is relying on file storage.
- initialize (this handles creating the directory, initializing the file and writing the brain parameters, which can remain as separate, private methods)
- record (is just a rename of WriteExperience)
- close (handles saving metadata)

awjuliani replied:
Makes a lot of sense. Will do.

mmattar commented on UnitySDK/Assets/ML-Agents/Scripts/DemonstrationStore.cs:
```diff
@@ -0,0 +1,103 @@
+
+            // Write AgentInfo to file.
+            var jsonInfo = JsonUtility.ToJson(info);
+            var writer = new StreamWriter(filePath, true);
```
Would it make sense to keep the writer object open from initialize to close?

mmattar commented on UnitySDK/Assets/ML-Agents/Scripts/DemonstrationStore.cs:
```diff
@@ -0,0 +1,103 @@
+    /// Demonstration meta-data.
+    /// </summary>
+    [System.Serializable]
+    public struct DemonstrationMetaData
```
Curious as to why this is a separate struct as opposed to instance variables for DemonstrationStore.

mmattar replied:
Makes sense - that would be a nice comment for the Demo struct.

mmattar commented on UnitySDK/Assets/ML-Agents/Scripts/DemonstrationStore.cs:
```diff
@@ -0,0 +1,103 @@
+        /// </summary>
+        public void CreateDirectory()
+        {
+            if (!Directory.Exists("Assets/Demonstrations"))
```
"Assets/Demonstrations" should be defined as a const var

mmattar commented on UnitySDK/Assets/ML-Agents/Scripts/DemonstrationStore.cs:
```diff
@@ -0,0 +1,103 @@
+        public int API;
+    }
+    
+    public class DemonstrationStore
```
Two design-focused questions:
- Why can't we support addition of experiences? I imagine a common workflow would be: record some experiences, attempt to train, realized it doesn't do so well and then go back to add more experiences?
- Given that the demo file will be consumed from Python, why not follow the Protobuf definitions as we have for the rest?

mmattar commented on UnitySDK/Assets/ML-Agents/Scripts/DemonstrationStore.cs:
```diff
@@ -0,0 +1,103 @@
+            {
+                numberEpisodes = 0,
+                numberExperiences = 0,
+                API = 1
```
We should have an API_VERSION const that is set to 1.

mmattar commented on UnitySDK/Assets/ML-Agents/Scripts/DemonstrationStore.cs:
```diff
@@ -0,0 +1,103 @@
+﻿using System.IO;
+using UnityEngine;
+
+
```
remove empty line

mmattar commented on UnitySDK/Assets/ML-Agents/Scripts/DemonstrationStore.cs:
```diff
@@ -0,0 +1,103 @@
+        }
+
+        /// <summary>
+        /// Writes brain parameters as json to file.
```
As much as possible, method-level comments should not expose implementation details (e.g. saving to file, JSON format)- it should be more of the what than how. This comment applies to a few method docs elsewhere.

xiaomaogy commented:
I am not sure if I am using the new feature correctly. I tried this with the

awjuliani commented:
@eshvk @vladimiroster @mmattar 

Thanks for all the feedback. New version now ready for another review. New outline is below:

## Specifics C#:
* `DemonstrationRecorder` (Class - Inherits from MonoBehavior) - Component which attaches to GameObject with Agent Component, contains `DemonstrationStore`. 
* `DemonstrationStore` (Class) contains all actual demonstration writing logic.
* `Demonstration` (ScriptableObject) contains metadata about recorded demonstration.
* `DemonstrationImporter` (ScriptedImporter) contains logic to load `.demo` file as Unity asset.
## Specifics Python:
* Function `load_demonstration` which takes file path of `.demo` file, and returns a filled buffer ready to use for training.
* Rename `BehavioralCloningTrainer` to `OnlineBCTrainer`.
* Create `OfflineBCTrainer` which uses `DemonstrationLoader` and `.demo` file for experience instead of environment.
* Both BC algorithms use `BCTrainer` base class.
## Demonstrations:
* File is of type `.demo`. Contains series of Protobuf objects.
* First 32 bytes dedicated to metadata proto object.
* Next chunk of data dedicated to `BrainPrameters` proto object.
* All following data dedicated to sequence of `AgentInfo` proto objects.

vincentpierre commented on UnitySDK/Assets/ML-Agents/Scripts/Utilities.cs:
```diff
@@ -0,0 +1,77 @@
+        /// If set to <c>true</c> the textures
+        /// will be converted to grayscale before being stored in the tensor.
+        /// </param>
+        public static float[,,,] BatchVisualObservations(
```
Maybe rename `TextureToFloatArray` ?

awjuliani commented:
@eshvk The issue with just using `.pb` is that the asset importer won't know that it is specifically a demonstration file. If other proto files are added to the project it will attempt to import them as well for example, and vise-versa for other Unity plugins which might use proto files.

vincentpierre commented:
How about `.pbrec` (protobuf recording), `.mlrec`(ml recording) or `.bcrec`(behavioral cloning recording) ?

vincentpierre commented on ml-agents/mlagents/trainers/demo_loader.py:
```diff
@@ -0,0 +1,144 @@
+
+logger = logging.getLogger("mlagents.trainers")
+
+
```
Should these line in a class called `DemonstrationLoader` and be made static methods rather than having unstructured methods ?

awjuliani replied:
@eshvk Suggested they be methods not connected to a class.

awjuliani commented:
> How about `.pbrec` (protobuf recording), `.mlrec`(ml recording) or `.bcrec`(behavioral cloning recording) ?

Maybe `.mldemo` ?
